Sources/Networking/HTTPClient/HTTPClient.swift

```diff
@@ -582,6 +606,95 @@ private extension HTTPClient {
+
+        // At this point, retryCount hasn't been incremented yet, so we'll need to do it early here
+        // to determine if another retry is appropriate.
+        let nextRetryCount = request.retryCount + 1
+
+        guard nextRetryCount <= self.retryBackoffIntervals.count else {
+            Logger.error(
+                NetworkStrings.api_request_failed_all_retries(
+                    httpMethod: request.method.httpMethod,
+                    path: request.path,
+                    retryCount: request.retryCount
+                )
+            )
+            return false
+        }
+
+        let retryBackoffInterval: TimeInterval = calculateRetryBackoffTime(
+            forResponse: httpURLResponse,
+            retryCount: nextRetryCount
+        )
+
+        Logger.debug(
+            NetworkStrings.api_request_queued_for_retry(
+                httpMethod: request.method.httpMethod,
+                path: request.path,
+                backoffInterval: retryBackoffInterval
+            )
+        )
+        self.operationDispatcher.dispatchOnWorkerThread(delay: .timeInterval(retryBackoffInterval)) {
+            let retriedRequest = request.retriedRequest()
+            self.state.modify {
+                $0.queuedRequests.insert(retriedRequest, at: 0)
+            }
+        }
+        return true
+    }
+
+    internal func shouldRetryRequest(withStatusCode statusCode: HTTPStatusCode) -> Bool {
+        return self.retriableStatusCodes.contains(statusCode)
+    }
+
+    internal func calculateRetryBackoffTime(
+        forResponse httpURLResponse: HTTPURLResponse,
+        retryCount: UInt
+    ) -> TimeInterval {
+        // Use the retry after value from the backend if present
+        if let retryAfterHeaderValue = httpURLResponse.allHeaderFields[ResponseHeader.retryAfter.rawValue] as? String {
+            if let retryAfterMS = Double(retryAfterHeaderValue) {
+                return TimeInterval(milliseconds: retryAfterMS)
+            }
+        }
+
+        // Otherwise, use a default value
+        return defaultExponentialBackoffTimeInterval(withRetryCount: retryCount)
+    }
+
+    /// Returns the exponential backoff time interval for a given retry count.
+    ///
+    /// This function calculates the backoff time interval for a given retry count based on a predefined
+    /// list of backoff intervals. The `retryCount` is not 0-indexed and should start with 1.
+    ///
+    /// - Parameter retryCount: The count of the retry attempt, starting from 1.
+    /// - Returns: The backoff time interval for the given retry count. If the retry count exceeds
+    ///   the predefined list of backoff intervals, it returns 0.
+    internal func defaultExponentialBackoffTimeInterval(withRetryCount retryCount: UInt) -> TimeInterval {
+        return retryCount - 1 < self.retryBackoffIntervals.count
+        ? self.retryBackoffIntervals[Int(max(retryCount - 1, 0))]
```

Comment:
Extremely very nit, but we could extract the max() for more clarity:

```
var backoffIntervalIndex = Int(max(retryCount - 1, 0))
return backoffInternalIndex < self.retryBackoffIntervals.count ? self.retryBackoffIntervals[backoffInternalIndex] : 0
```

Reply:
Love this! Will make the change.